src/chapters/typescript/functions.rst

```diff
@@ -0,0 +1,115 @@
+We can make anonymous functions, give it a different number of parameters and so on.
+However, when working with functions in TypeScript, we have to keep *type* in mind.
+
+Declaring Functions
+-------------------
+
+Named Functions
+^^^^^^^^^^^^^^^
+
+Let's take a look at a TypeScript function declaration.
+
+.. sourcecode:: typescript
+
+   function myFunction(a: number): number {
+      return a*2;
+   }
+
+Here you can see that we provided the type of the parameters and the type of the value that is returned.
+If the function doesn't have a return value, then you use ``void`` for its type.
+
+.. note::
+
+   While in many cases, it is possible to determine the return type during compilation, if you know the return type, you should give it. 
+
+Anonymous Functions
+^^^^^^^^^^^^^^^^^^^
+
+For an anonymous function, you still need to provide types for the value returned and the parameters.
+
+.. sourcecode:: typescript
+
+   let myFunction = function(a: number): number { return a*2; };
+
+The above code looks very much like an anonymous function declaration with the types included!
+However, TypeScript has arrow notation and here we use the **fat arrow**, ``=>``.
+
+.. sourcecode:: typescript
+
+   let myFunction: (a: number) => number = function(a: number): number { return a*2; };
+
+In TypeScript, the keywords that come after the colon are the type. The ``=>`` operator then indicates the return type of the function.
+
+Optional Parameter
+------------------
+
+In JavaScript, you can declare a function with 5 parameters and only give it 2 when it comes time to use it.
+In TypeScript, that is not the case. When you declare a function with 5 parameters, you have to give the function 5 arguments when calling it.
+However, in TypeScript, you can make parameters optional. This means that when you are calling the function, you can leave off the optional parameter.
+
+To denote a parameter as optional, you can use ``?`` notation. Any parameters that are optional must follow the required parameters.
+
+.. admonition:: Example
+
+   .. sourcecode:: typescript
+
+      function myFunction(a: number, b?:number): number {
+         return a+b;
+      }
+
+      console.log(myFunction(1,2));
+      console.log(myFunction(1));
+      console.log(myFunction(3,5));
+      console.log(myFunction(3));
+
+   ::
+
+      3
+      1
+      8
+      3
+
+In this example, the ``myFunction`` function has two parameters, ``a`` and ``b``. ``a`` is required and ``b`` is optional. When only one argument is provided to the function, then the value of ``a`` is returned.
+When both arguments are provided, then the sum of the 2 arguments is returned.
+
+Another way to deal with a parameter you need to be optional is to give it a default value.
+
+Let's say that in the example above, we wanted to give ``b`` a default value of 5. That way, if no argument is supplied for ``b``, then the value returned is that of ``a+5``.
```

Comment:
i think this example can be stronger if the return value when the optional parameter is left off is not the same as the first parameter. does that make sense?

Reply:
oops! this note would apply to the previous example, not the case where b has a default value

Reply:
Added another number to the sum so that the example will reinforce the fact that the parameter is optional!